File: geekbang/algorithm-models/7.2- Dijkstra 求最短路代码模板/dijkstra.py
# Python
# https://www.acwing.com/problem/content/description/852/
from heapq import *


if __name__ == "__main__":
    n, m = map(int,input().split())

    ver = [[] for i in range(n + 1)] # 0~n
    edge = [[] for i in range(n + 1)] # 0~n
    dist = [1e9] * (n + 1)
    v = [False] * (n + 1)

    # 出边数组建图
    for i in range(m):
        x, y, z = map(int,input().split())
        ver[x].append(y)  # 另一端点
        edge[x].append(z) # 边权


    heap = []
    heappush(heap, (0, 1)) # (距离, 点)
    dist[1] = 0


    # Dijkstra 算法
    while heap:
        distance, x = heappop(heap)
        if v[x]:
            continue
        v[x] = True


        for i in range(len(ver[x])):
            y, z = ver[x][i], edge[x][i]
            if dist[y] > dist[x] + z:
                dist[y] = dist[x] + z
                heappush(heap, (dist[y], y))


    print(dist[n] if dist[n] != 1e9 else -1)

File: geekbang/algorithm-models/7.2- Dijkstra 求最短路代码模板/dijkstra.java
// Java
// https://www.acwing.com/problem/content/description/852/
import java.io.*;
import java.util.*;


public class Main {
    public static void main(String args[]) throws Exception {
        Scanner input = new Scanner(System.in);
        int n = input.nextInt();
        int m = input.nextInt();
        
        // 模板：出边数组初始化
        // 初态：[[], [], ... []]
        List<List<Integer>> ver = new ArrayList<List<Integer>>(); // 另一端点
        List<List<Integer>> edge = new ArrayList<List<Integer>>(); // 边权
        boolean[] v = new boolean[n + 1];
        int[] dist = new int[n + 1];
        for (int i = 0; i <= n; i++) {
            ver.add(new ArrayList<Integer>());
            edge.add(new ArrayList<Integer>());
            v[i] = false;
            dist[i] = (int)1e9;
        }


        for (int i = 1; i <= m; i++) {
            int x = input.nextInt();
            int y = input.nextInt();
            int z = input.nextInt();
            // 出边数组 addEdge 模板
            ver.get(x).add(y);
            edge.get(x).add(z);
        }
        
        // Dijkstra算法
        dist[1] = 0;
        // 堆，每个结点是长度为2的数组 [点，dist]
        PriorityQueue<int[]> q = new PriorityQueue<>((a,b) -> {return a[1] - b[1];});
        q.offer(new int[]{1, 0});
        while(!q.isEmpty()){
            int[] top = q.poll();
            int x = top[0];
            if (v[x]) continue;
            v[x] = true;
            for (int i = 0; i < ver.get(x).size(); i++) {
                int y = ver.get(x).get(i);
                int z = edge.get(x).get(i);
                if (dist[y] > dist[x] + z) {
                    dist[y] = dist[x] + z;
                    q.offer(new int[]{y, dist[y]});
                }
            }
        }
        System.out.println(dist[n] == 1e9 ? -1 : dist[n]);
    }
}
